README
AVbin
=====

Home:       http://avbin.github.com
Maintainer: Nathan Stocks <nathan.stocks@gmail.com>


End-User Installation
---------------------

End-users should use the installers provided in the binary release.  That is one
of the benefits of a binary distribution, after all...


Developing
----------

AVbin is a C library following standard conventions.

Clone the git repository:

  git clone https://github.com/AVbin/AVbin.git avbin

See the header file ``include/avbin.h`` and the online API documentation at
http://avbin.github.com/AVbin/Docs.html


Building
--------

If you are interested in testing, contributing code, or otherwise helping out,
we encourage you to join the Pyglet mailing list (AVbin was originally created
for the Pyglet project).  See http://pyglet.org/contribute.html

AVbin is designed to be used with a particular version of Libav.  This version
of Libav is included as a git submodule of the AVbin project under the ``libav``
directory.  Note that this version of Libav may contain AVbin-specific patches,
though we try to get all patches accepted upstream.

Use the included `build.sh` script to compile AVbin.  This takes care of
configuring Libav correctly, compiling it, then linking it with the AVbin
sources.  The final AVbin libraries are placed in the ``dist`` directory.

To build on your platform, use the included ``build.sh`` script.  See
``build.sh --help`` for usage information.

If you are planning on distributing the self-built binary to anyone other than
yourself, please create a "PRERELEASE" file in the root of the project directory
containing a version modifier such as "-my-version".
get_avbin_info()->version_string would then return "10-my-version" for a custom
build of AVbin 10, while get_avbin_info()->version will still return the integer
10.


Building for OS X
-----------------

Tested on OS X Lion (10.7) and Mountain Lion (10.8)

- Install Xcode (from Apple's App Store)
- Open Xcode, and go to ``Preferences > Downloads > Components`` and install the
  Command Line Tools package.
- Install yasm (You can install HomeBrew from http://mxcl.github.io/homebrew/
  and then run 'brew install yasm' to install yasm)
- Obtain and install Auxiliary Tools for Xcode (from developer.apple.com > 
  Mac Dev Center > Developer Downloads)
- Run ./build.sh [desired options and target]


Building for Linux
------------------

Tested on a default install of Ubuntu 12.04 LTS (both 32- and 64-bit installs).

- Install yasm: sudo apt-get install yasm
- Install libbz2-dev: sudo apt-get install libbz2-dev
- (If you want to make the binary installer)
  Install makeself: sudo apt-get install makeself
- Run ./build.sh [desired options and targets]


Building for Win64 or Win32 (crosscompiling on Linux)
-----------------------------------------------------

Tested on Ubuntu 12.04 LTS 64-bit.

- Install MinGW-64: sudo apt-get install mingw-64
- (If you want to make the binary installer)
  Install nsis: sudo apt-get install nsis nsis-doc
- Cross-compile zlib and install it:

  wget http://zlib.net/zlib-1.2.7.tar.gz
  tar zxvf zlib-1.2.7.tar.gz
  cd zlib-1.2.7

  # For win32...
  make PREFIX=i686-w64-mingw32- -f win32/Makefile.gcc
  sudo make PREFIX=i686-w64-mingw32- \
    LIBRARY_PATH=/usr/i686-w64-mingw32/lib \
    INCLUDE_PATH=/usr/i686-w64-mingw32/include/ \
    BINARY_PATH=/dev/null \
    -f win32/Makefile.gcc install
  make -f win32/Makefile.gcc clean

  # For win64...
  make PREFIX=x86_64-w64-mingw32- -f win32/Makefile.gcc
  sudo make PREFIX=x86_64-w64-mingw32- \
    LIBRARY_PATH=/usr/x86_64-w64-mingw32/lib \
    INCLUDE_PATH=/usr/x86_64-w64-mingw32/include/ \
    BINARY_PATH=/dev/null \
    -f win32/Makefile.gcc install

- Cross-compile bzlib and install it

  wget http://bzip.org/1.0.6/bzip2-1.0.6.tar.gz
  tar zxvf bzip2-1.0.6.tar.gz
  cd bzip2-1.0.6/

  # For win32...
  make CC=i686-w64-mingw32-gcc AR=i686-w64-mingw32-ar \
    RANLIB=i686-w64-mingw32-ranlib PREFIX=/usr/i686-w64-mingw32 libbz2.a
  sudo cp libbz2.a /usr/i686-w64-mingw32/lib/
  sudo cp bzlib.h /usr/i686-w64-mingw32/include/
  make clean

  # For win64...
  make CC=x86_64-w64-mingw32-gcc AR=x86_64-w64-mingw32-ar \
    RANLIB=x86_64-w64-mingw32-ranlib PREFIX=/usr/x86_64-w64-mingw32 libbz2.a
  sudo cp libbz2.a /usr/x86_64-w64-mingw32/lib/
  sudo cp bzlib.h /usr/x86_64-w64-mingw32/include/

- (For creating dist zipfile) Install 7-Zip: sudo apt-get install p7zip-full

- Run ./build.sh [desired options and targets]


Caveats
-------

AVbin is currently known to function on the following platforms

* Linux x86 (32- and 64-bit)

* Mac OS X 10.6 - 10.7 (32- and 64-bit)

* Windows XP (32-bit)

* Windows 7 (64-bit)


Creating a Distribution
-----------------------

- Using the same value for ``platform`` as the build you just did, run:
  ./dist.sh [platform]


Creating Documentation
----------------------

- Install Doxygen from http://www.stack.nl/~dimitri/doxygen/

- Open/Edit ``Doxyfile`` to your liking and generate it with Doxygen.  The GUI
  frontend seems to work just fine.

Installation and Usage
----------------------

Run the binary installer, or place the resulting avbin.so, avbin.dylib or
avbin.dll from the ``dist`` directory into the appropriate system directory.

The AVbin dynamic library exports all of Libav's functions from libavcodec,
libavutil, libavformat, and libswscale.  It also exports some higher-level
functions which have a fixed ABI (they will not change in incompatible ways
in future releases), documented in ``include/avbin.h``.


License
-------

Due to the linkage between AVbin and Libav, AVbin must be licensed under the
LGPL or GPL.  Currently all GPL features of the Libav configuration are
disabled, and AVbin is licensed under the LGPL.

You should see the accompanying COPYING.LESSER file for details.  In summary,
you must note the usage of AVbin within the documentation of your application.
If you make changes to either library, you must either contribute said changes
upstream or include the sources of these changes within your distributed
application.


Contributions
-------------

Nathan Stocks      - Project maintainer
Alex Holkner       - Original author, original maintainer
Micah Richert      - Contributed original macosx-x86-64 and win64 build scripts
Anatoly Techtonik  - C++ integration
Jernej Virag       - C code additions and updates, build script updates
Jeremy Lujan       - OS X install script patches.
Tim Sheerman-Chase - avbin_seek_file() patches and Visual Studio 2010 compatibility


CHANGELOG
AVbin 11

- Updated the contributor list at the bottom of the README file
- Fixed all the copyright notices and license headers that needed fixing.
- Removed the sample format AVBIN_SAMPLE_FORMAT_S24 which was removed upstream
  and deprecated in AVbin 8.
- We now pad our input buffers internally as the Libav API specifies we should.
- Updated Livab to upstream master branch commit
  720a1de52fb4abd8e0541cdcc7d57cb0044c7d3a
- Fixed our AVFrame handling to conform with new upstream requirements which
  allows us to pull in an updated version of Libav.
- Updated the OS X section of the readme to reflect OS X 10.8 and Brew.
- Removed AVBIN_SAMPLE_FORMAT_S24 that was deprecated in AVbin 8.
- Deprecated avbin_get_audio_buffer_size() since it is both entirely useless
  (it has always returned the constant 192000) and has been removed upstream.
  It will be permanently removed in AVbin 13.
- Installer filenames now include prerelease version information, if
  applicable.
- Statically linking libbz2, since it's getting crazy hard to make it
  dynamically work for all the crazy ways different Linux distros want their
  own version of it.
- av_stream_info() now reports the correct sample rate for 8-bit files (a bug
  was causing 8-bit files to always report a sample rate of 0)
- avbin_get_ffmpeg_revision() deprecation compiler warning works in Visual
  Studio 2010.  Thanks to Tim Sheerman-Chase.
- av_seek_file() is now more accurate and stable.  Thanks to Tim Sheerman-Chase.
- If av_seek_file() fails to seek, it now properly returns AVBIN_RESULT_ERROR.
- Removed "frame_rate" feature (reflecting upstream removal).
- Fixed a crash that could happen in the OS X installer if /usr/local/lib
  didn't already exist.  Thanks to Jeremy Lujan.
- Fixed the uninstaller for win32

AVbin 10

- Updated to Libav trunk as of 2012-08-12 (which finally fixed the .ogg decoding
  problems on all platforms!).

- Added the avbin_get_info() function and removed the avbin_get_libav_commit()
  and avbin_get_libav_version() functions that I tried out in AVbin 9, which I
  am now pretending never got released.

  avbin_get_info() returns an _AVbinInfo struct (typedef AVbinInfo) with the
  following members (all strings except where indicated):
  - version                 > integer, same as avbin_get_version()
  - version_string          > potentially includes prerelease info,
                              for example: "10-beta1"
  - build_date              > %Y-%m-%d %H:%M:%S %z
  - repo                    > Git URL
  - commit                  > Git commit hash
  - backend                 > "libav" or "ffmpeg"
  - backend_version_string  > Most recent backend version tag or trunk and date
                              i.e. "v0.8.3" or "trunk (Wed Aug 22 19:58:46 2012 -0400)"
  - backend_repo            > Backend git URL
  - backend_commit          > Backend git commit hash

- Added AVbinOptions typedef (struct _AVbinOptions) for use with the new
  avbin_init_options() function.

- Added avbin_init_options() to handle initialization options.  avbin_init() now
  creates an AVbinOptions with default settings and calls avbin_init_options()
  under the hood.  Future versions of AVbin may have new AVbinOptions* structs
  that you can pass to avbin_init_options().  Currently, the only option is how
  many threads you would like avbin to use for decoding (defaults to 1, using
  0 [autodetect] is recommended).

- Updated avbin_have_features() with new features to check for.
  "options" > presence of avbin_init_options(), AVbinOptions, and multithreading
  "info"    > presence of avbin_get_info(), AVbinInfo

- Adapted the underlying C code implementation to upstream API changes, as well
  as switching to more portable data types across the board.  This work was
  almost entirely pulled from Jernej Virag's fork, so kudos to him!

- Implemented multithreading support inspired by Jernej's work.  If you want
  the library to use multiple threads, please see avbin_init_options() in the
  documentation.

- Overhauled all configure options and Makefile commands so that all six
  supported platforms have very similar builds now.  Bz2 and Gzip compression is
  enabled on all platforms now, for example.  Every platform also had to have
  configure and make options adjusted to compensate for new versions of Xcode
  and OS X, new versions of Linux, and the new version of Libav.

- Overhauled the build script.  All build errors now have descriptive messages.
  The options may now be specified at any argument position.  The --fast option
  now compiles with more processes at once.  The --clean option now does a
  better job cleaning.  The libav submodule is automatically updated before
  building.

- Overhauled the dist script, which now generates binary installers for all
  platforms

- The OS X version is now a universal library containing both 32-bit and 64-bit
  code.

- The OS X release is now a package installer, instead of a tarball.

- The Linux release is now a compressed binary installer instead of a tarball.

- The Windows release is now an executable installer instead of a zipfile.

- All installers now present and require acceptance of the license.

- Updated the Makefile for the avbin_dump example so that it compiles on Linux

- Updated the avbin_dump example:
  - Command-line arguments are now processed position-independently
  - The new help argument (-h, --help) displays usage
  - The new verbose argument (-v, --verbose) enables printing out information
    about each packet decoded from a media file.  (Previously you had to modify
    the source code and recompile to enable this).
  - The verbose output now includes timestamps (in microseconds).

- Updated the online API documentation: http://avbin.github.com/docs/

- Ran the repository through `git filter-branch` and eliminated all of the
  ffmpeg source code, which shrinks the repo size to 528K (much better than

build.sh
#!/bin/bash
#
# build.sh
# Copyright 2007 Alex Holkner
# Copyright 2011-2013 AVbin Team
#
# This file is part of AVbin.
#
# AVbin is free software; you can redistribute it and/or modify it
# under the terms of the GNU Lesser General Public License as
# published by the Free Software Foundation; either version 3 of
# the License, or (at your option) any later version.
#
# AVbin is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this program.  If not, see
# <http://www.gnu.org/licenses/>.

# Either "libav" or "ffmpeg"
BACKEND=libav

# Directory holding backend source code.
BACKEND_DIR=$BACKEND

# Make sure the git submodule is initiated and updated (if blank)
if ! ls $BACKEND_DIR/Makefile >/dev/null 2> /dev/null ; then
    git submodule init --quiet
    git submodule update
fi


AVBIN_VERSION=`cat VERSION`
AVBIN_PRERELEASE=`cat PRERELEASE 2>/dev/null`
AVBIN_VERSION_STRING="\"$AVBIN_VERSION$AVBIN_PRERELEASE\""
AVBIN_BUILD_DATE="\"`date +\"%Y-%m-%d %H:%M:%S %z\"`\""
AVBIN_REPO="\"`git remote show -n origin | grep Fetch | cut -b 14-`\""
AVBIN_COMMIT="\"`git rev-parse HEAD`\""

# Backend information
AVBIN_BACKEND="\"$BACKEND\""
pushd $BACKEND_DIR > /dev/null
AVBIN_BACKEND_VERSION_STRING="\"`git describe --abbrev=0 --tags`\""
if [ $AVBIN_BACKEND_VERSION_STRING == "\"v0.8\"" ] ; then
    TRUNK_DATE=`git log -1 --pretty=format:%cd`
    AVBIN_BACKEND_VERSION_STRING="\"trunk ($TRUNK_DATE)\""
fi
AVBIN_BACKEND_REPO="\"`git remote show -n origin | grep Fetch | cut -b 14-`\""
AVBIN_BACKEND_COMMIT="\"`git rev-parse HEAD`\""
popd > /dev/null

fail() {
    echo "AVbin: Fatal error: $1"
    exit 1
}

clean_backend() {
    pushd $BACKEND_DIR > /dev/null
    echo -n "Cleaning up..."
    make clean 2> /dev/null
    make distclean 2> /dev/null
    echo "done"
    find . -name '*.d' -exec rm -f '{}' ';'
    find . -name '*.pc' -exec rm -f '{}' ';'
    rm -f config.log config.err config.h config.mak .config .version
    popd > /dev/null
}

build_backend() {
    config=`pwd`/$PLATFORM.configure
    common=`pwd`/common.configure

    if [ ! $REBUILD ]; then
        clean_backend
    fi

    pushd $BACKEND_DIR > /dev/null

    # If we're not rebuilding, then we need to configure Backend
    if [ ! $REBUILD ]; then
        case $OSX_VERSION in
            "10.6") SDKPATH="\/Developer\/SDKs\/MacOSX10.6.sdk" ;;
            "10.7") SDKPATH="\/Applications\/Xcode.app\/Contents\/Developer\/Platforms\/MacOSX.platform\/Developer\/SDKs\/MacOSX10.6.sdk" ;;
            *)      SDKPATH="" ;;
        esac

        cat $config $common | egrep -v '^#' | sed s/%%SDKPATH%%/$SDKPATH/g | xargs ./configure || fail "Failed configuring backend."
    fi

    # Remove -Werror options from config.mak that break builds on some platforms
    cat config.mak | sed -e s/-Werror=implicit-function-declaration//g | sed -e s/-Werror=missing-prototypes//g > config.mak2
    mv config.mak2 config.mak

    # Actually build Backend
    make $FAST || fail "Failed to build backend."
    popd
}

build_avbin() {
    # For avbin.c ...
    export AVBIN_VERSION
    export AVBIN_VERSION_STRING
    export AVBIN_BUILD_DATE
    export AVBIN_REPO
    export AVBIN_COMMIT
    export AVBIN_BACKEND
    export AVBIN_BACKEND_VERSION_STRING
    export AVBIN_BACKEND_REPO
    export AVBIN_BACKEND_COMMIT

    # For the Makefile ...
    export PLATFORM
    export BACKEND_DIR

    if [ ! $REBUILD ]; then
        make clean
    fi

    make || fail "Failed to build AVbin."
}

build_macosx_universal() {
    if [ ! -e dist/macosx-x86-32/libavbin.$AVBIN_VERSION.dylib ]; then
        PLATFORM=macosx-x86-32
        build_backend
        build_avbin
    fi

    if [ ! -e dist/macosx-x86-64/libavbin.$AVBIN_VERSION.dylib ]; then
        PLATFORM=macosx-x86-64
        build_backend
        build_avbin
    fi

    mkdir -p dist/macosx-universal
    lipo -create \
        -output dist/macosx-universal/libavbin.$AVBIN_VERSION.dylib \
        dist/macosx-x86-32/libavbin.$AVBIN_VERSION.dylib \
        dist/macosx-x86-64/libavbin.$AVBIN_VERSION.dylib || fail "Failed to create universal shared library."
}

die_usage() {
    echo "Usage: ./build.sh [options] <platform> [<platform> [<platform> ...]]"
    echo
    echo "Options"
    echo "  --clean     Don't build, just clean up all generated files and directories."
    echo "  --fast      Use 'make -j4' when compiling"
    echo "  --help      Display this help text."
    echo "  --rebuild   Don't reconfigure, just run make again."
    echo
    echo "Supported platforms:"
    echo "  linux-x86-32"
    echo "  linux-x86-64"
    echo "  macosx-x86-32"
    echo "  macosx-x86-64"
    echo "  macosx-universal (builds all supported Mac OS X architectures into one library)"
    echo "  win32"
    echo "  win64"
    exit 1
}

for arg in $* ; do
    case $arg in
        "--fast")
            FAST="-j4" ;;
        "--help")
            die_usage ;;
        "--rebuild")
            REBUILD=1;;
        "--clean")
            clean_backend
            rm -rf dist
            rm -rf build
            rm -f example/avbin_dump
            rm -f example/minimal
            exit
            ;;
        *)
            if [ "$platforms" == "" ]; then
                platforms=$arg
            else
                platforms="$platforms $arg"
            fi
            ;;
    esac
done;

if [ "$platforms" == "" ]; then
    die_usage
fi

for PLATFORM in $platforms; do
    case $PLATFORM in
        "macosx-universal")
            OSX_VERSION=`/usr/bin/sw_vers -productVersion | cut -b 1-4`
            build_macosx_universal
            ;;
        "macosx-x86-32" | "macosx-x86-64")
            OSX_VERSION=`/usr/bin/sw_vers -productVersion | cut -b 1-4`
            build_backend
            build_avbin
            ;;
        "linux-x86-32" | "linux-x86-64" | "win32" | "win64")
            build_backend
            build_avbin
            ;;
        *)
            echo "Unrecognized platform: $PLATFORM.  Try ./build.sh --help" && exit 3
            ;;
    esac
done

echo "Build successful."
